File: c/2-motor/perf.c
#include <bits/time.h>
#include <inttypes.h>
#include <stdio.h>
#include <time.h>

#include "perf.h"

int perf_counter_init(perf_counter_t *self, const char *name) {
  int res;

  struct timespec resolution;
  res = clock_getres(CLOCK_THREAD_CPUTIME_ID, &resolution);
  if (res != 0) {
    return res;
  }

  printf(
      "Performance counter %s, cpu resolution: %ld s %ld ns\n", name,
      resolution.tv_sec, resolution.tv_nsec
  );

  *self = (perf_counter_t){
      .name = name,
      .total_time = {.tv_sec = 0, .tv_nsec = 0},
      .sample_count = 0,
  };

  return 0;
}

perf_mark_t perf_mark() {
  struct timespec mark;
  int res = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &mark);
  if (res != 0)
    return (struct timespec){.tv_sec = 0, .tv_nsec = 0};
  else
    return mark;
}
void perf_counter_add_sample(perf_counter_t *self, perf_mark_t start) {
  const perf_mark_t end = perf_mark();
  const struct timespec diff = {
      .tv_sec = end.tv_sec - start.tv_sec,
      .tv_nsec = end.tv_nsec - start.tv_nsec,
  };

  self->total_time = (struct timespec){
      .tv_sec = self->total_time.tv_sec + diff.tv_sec,
      .tv_nsec = self->total_time.tv_nsec + diff.tv_nsec,
  };
  self->sample_count += 1;
}

void perf_counter_report(perf_counter_t *const self) {
  const double time_us =
      ((double)self->total_time.tv_nsec / 1e3 + self->total_time.tv_sec * 1e6) /
      self->sample_count;
  printf(
      "Performance counter %s: %.3f us (%d sampl.)\n", self->name, time_us,
      self->sample_count
  );
}
void perf_counter_reset(perf_counter_t *self) {
  self->total_time = (struct timespec){.tv_sec = 0, .tv_nsec = 0};
  self->sample_count = 0;
}

File: zig/2-motor/perf.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const c = @import("c.zig");

pub const Marker = struct {
    time_ns: u64,

    pub fn now() Marker {
        var time: c.struct_timespec = undefined;
        const res = c.clock_gettime(c.CLOCK_THREAD_CPUTIME_ID, &time);
        const time_ns = if (res != 0) 0 else ns_from_timespec(&time);
        return Marker{ .time_ns = time_ns };
    }
};

pub const Counter = struct {
    const Self = @This();

    name: []const u8,
    samples_ns: std.ArrayList(u32),

    // Panics instead of errors, not to influence CCN metric
    pub fn init(allocator: Allocator, name: []const u8, length: usize) Self {
        var resolution: c.struct_timespec = undefined;
        const res = c.clock_getres(c.CLOCK_THREAD_CPUTIME_ID, &resolution);
        if (res != 0) {
            const err = std.posix.unexpectedErrno(std.posix.errno(res));
            std.debug.panic("Counter clock resolution: {}", .{err});
        }

        std.log.info(
            "Performance counter {s}, cpu resolution: {} ns",
            .{ name, ns_from_timespec(&resolution) },
        );

        const samples_ns = std.ArrayList(u32).initCapacity(allocator, length) catch |err| {
            std.debug.panic("Counter buffer init: {}", .{err});
        };

        return .{
            .name = name,
            .samples_ns = samples_ns,
        };
    }

    pub fn deinit(self: *const Counter) void {
        self.samples_ns.deinit();
    }

    pub fn add_sample(self: *Self, start: Marker) void {
        const end = Marker.now();
        const diff = end.time_ns - start.time_ns;

        if (self.samples_ns.items.len >= self.samples_ns.capacity) {
            std.log.err("perf.Counter.add_sample: buffer is full", .{});
            return;
        }

        const sample = self.samples_ns.addOneAssumeCapacity();
        sample.* = @truncate(diff);
    }

    pub fn report(self: *const Self) void {
        std.log.info(
            "Performance counter {s}: {}",
            .{ self.name, SampleFormatter{ .data = self.samples_ns.items } },
        );
    }

    const SampleFormatter = std.fmt.Formatter(formatSamples);

    fn formatSamples(
        data: []const u32,
        comptime _: []const u8,
        _: std.fmt.FormatOptions,
        writer: anytype,
    ) !void {
        try writer.writeAll("[");
        for (data, 0..) |sample, i| {
            try std.fmt.format(writer, "{d}", .{sample / 1000});
            if (i < data.len - 1)
                try writer.writeAll(",");
        }
        try writer.writeAll("] us");
    }

    pub fn reset(self: *Self) void {
        self.samples_ns.clearRetainingCapacity();
    }
};

fn ns_from_timespec(timespec: *const c.struct_timespec) u64 {
    const tv_sec: u64 = @intCast(timespec.tv_sec);
    const tv_nsec: u64 = @intCast(timespec.tv_nsec);
    return tv_nsec + tv_sec * std.time.ns_per_s;
}
